share/man/man7/ffs.7
.\" Copyright (c) 2001 Networks Associates Technology, Inc.
.\" All rights reserved.
.\"
.\" This software was developed for the FreeBSD Project by Chris
.\" Costello at Safeport Network Services and NAI Labs, the Security
.\" Research Division of Network Associates, Inc. under DARPA/SPAWAR
.\" contract N66001-01-C-8035 ("CBOSS"), as part of the DARPA CHATS
.\" research program.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: head/share/man/man7/ffs.7 148337 2005-07-23 15:09:27Z markus $
.\"
.Dd December 26, 2001
.Dt FFS 7
.Os
.Sh NAME
.Nm ffs
.Nd Berkeley fast file system
.Sh SYNOPSIS
In the kernel configuration file:
.Cd "options FFS"
.Cd "options QUOTA"
.Cd "options SOFTUPDATES"
.Cd "options SUIDDIR"
.Cd "options UFS_ACL"
.Cd "options UFS_DIRHASH"
.Cd "options UFS_EXTATTR"
.Cd "options UFS_EXTATTR_AUTOSTART"
.Pp
In
.Xr fstab 5 :
.Bd -literal -compact
/dev/disk0a	/mnt ufs rw 1 1
.Ed
.Sh DESCRIPTION
The Berkeley fast file system
provides facilities to store file system data onto a disk device.
.Nm
has been optimized over the years
for speed and reliability
and is the default
.Fx
file system.
.Ss Quotas
.Bl -tag -width 2n
.It Cd "options QUOTA"
This option allows system administrators
to set limits on disk usage
on a per-user basis.
Quotas can be used only on file systems
mounted with the
.Cm quota
option;
see
.Xr quota 1
and
.Xr edquota 8 .
.El
.Ss Soft Updates
.Bl -tag -width 2n
.It Cd "options SOFTUPDATES"
The soft updates feature tracks writes to the disk
and enforces metadata update dependencies
(e.g., updating free block maps)
to ensure that the file system remains consistent.
.Pp
To enable soft updates on an
.Em unmounted
file system, use the following command:
.Pp
.D1 Nm tunefs Fl n Cm enable Ar fs
.Pp
.Ar fs
can be either a mount point listed in
.Xr fstab 5
(e.g.,
.Pa /usr ) ,
or a disk device
(e.g.,
.Pa /dev/da0a ) .
.El
.Ss File Ownership Inheritance
.Bl -tag -width 2n
.It Cd "options SUIDDIR"
For use in file sharing environments
on networks including
.Tn "Microsoft Windows"
and
.Tn "Apple Macintosh"
computers,
this option allows files on file systems
mounted with the
.Cm suiddir
option
to inherit the ownership of its directory,
i.e.,
.Dq "if it's my directory, it must be my file."
.El
.Ss Access Control Lists
.Bl -tag -width 2n
.It Cd "options UFS_ACL"
Access control lists allow the association of
fine-grained discretionary access control information
with files and directories.
This option requires the presence of the
.Dv UFS_EXTATTR
option, and it is recommended that
.Dv UFS_EXTATTR_AUTOSTART
is included as well,
so that ACLs are enabled atomically upon mounting the file system.
.El
.Pp
In order to enable support for ACLs,
two extended attributes must be available in the
.Dv EXTATTR_NAMESPACE_SYSTEM
namespace:
.Pa posix1e.acl_access ,
which holds the access ACL,
and
.Pa posix1e.acl_default ,
which holds the default ACL for directories.
If you are using file system extended attributes,
the following commands may be used to
allocate space for and create the necessary EA backing files
for ACLs in the root of each file system.
In these examples, the root file system is used;
see
.Sx "Extended Attributes"
for more details.
.Bd -literal -offset indent
mkdir -p /.attribute/system
cd /.attribute/system
extattrctl initattr -p / 388 posix1e.acl_access
extattrctl initattr -p / 388 posix1e.acl_default
.Ed
.Pp
On the next mount of the root file system,
the attributes will be automatically started
(if
.Dv UFS_EXTATTR_AUTOSTART
is included in the kernel configuration),
and ACLs will be enabled.
.Ss Directory Hashing
.Bl -tag -width 2n
.It Cd "options UFS_DIRHASH"
Implements a hash-based lookup scheme for directories
in order to speed up accesses to very large directories.
.El
.Ss Extended Attributes
.Bl -tag -width 2n
.It Cd "options UFS_EXTATTR"
Extended attributes allow the association of
additional arbitrary metadata with files and directories,
which can be assigned and retrieved from userland
as well as from within the kernel; see
.Xr extattrctl 8 .
.It Cd "options UFS_EXTATTR_AUTOSTART"
If this option is defined,
.Nm
will search for a
.Pa .attribute
subdirectory of the file system root during the mount operation.
If found, extended attribute support will be
automatically started for that file system.
.El
.Pp
The following
.Xr sysctl 8
MIBs are defined for use with
.Nm :
.Bl -hang -width ".Va vfs.ffs.doreallocblk"
.It Va vfs.ffs.doasyncfree
Asynchronously write out modified i-node and indirect blocks
upon reallocating file system blocks to be contiguous.
(Default: 1.)
.It Va vfs.ffs.doreallocblks
Enable support for the rearrangement of blocks
to be contiguous.
(Default: 1.)
.El
.Sh SEE ALSO
.Xr quota 1 ,
.Xr acl 3 ,
.Xr extattr 3 ,
.Xr edquota 8 ,
.Xr extattrctl 8 ,
.Xr sysctl 8
.Rs
.%A M. McKusick
.%A W. Joy
.%A S. Leffler
.%A R. Fabry
.%D August 1984
.%T "A Fast File System for UNIX"
.%J "ACM Transactions on Computer Systems"
.%N 2
.%V 3
.%P 181-197
.Re
.Rs
.%A M. McKusick
.%D June 2000
.%T "Soft Updates: A Technique for Eliminating Most Synchronous Writes in the Fast Filesystem"
.%J "Proceedings of the Freenix Track at the 1999 Usenix Annual Technical Conference"
.%P 71-84
.Re


sys/libkern/arm/ffs.S
/*	$NetBSD: ffs.S,v 1.3 2003/04/05 23:27:15 bjh21 Exp $	*/
/*-
 * Copyright (c) 2001 Christopher Gilbert
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the company nor the name of the author may be used to
 *    endorse or promote products derived from this software without specific
 *    prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <machine/asm.h>
__FBSDID("$FreeBSD: head/sys/libkern/arm/ffs.S 172615 2007-10-13 12:05:36Z cognet $");


/*
 * ffs - find first set bit, this algorithm isolates the first set
 * bit, then multiplies the number by 0x0450fbaf which leaves the top
 * 6 bits as an index into the table.  This algorithm should be a win
 * over the checking each bit in turn as per the C compiled version.
 *
 * under ARMv5 there's an instruction called CLZ (count leading Zero's) that
 * could be used
 *
 * This is the ffs algorithm devised by d.seal and posted to comp.sys.arm on
 * 16 Feb 1994.
 */

ENTRY(ffs)

	/* Standard trick to isolate bottom bit in r0 or 0 if r0 = 0 on entry */
 	rsb     r1, r0, #0
 	ands    r0, r0, r1
#ifndef _ARM_ARCH_5
	
	/*
	 * now r0 has at most one set bit, call this X
	 * if X = 0, all further instructions are skipped
	 */
	adrne   r2, .L_ffs_table
	orrne   r0, r0, r0, lsl #4  /* r0 = X * 0x11 */ 
	orrne   r0, r0, r0, lsl #6  /* r0 = X * 0x451 */
	rsbne   r0, r0, r0, lsl #16 /* r0 = X * 0x0450fbaf */
              
	/* now lookup in table indexed on top 6 bits of r0 */
	ldrneb  r0, [ r2, r0, lsr #26 ]

	RET
.text;
.type .L_ffs_table, _ASM_TYPE_OBJECT;
.L_ffs_table:
/*               0   1   2   3   4   5   6   7           */
	.byte	 0,  1,  2, 13,  3,  7,  0, 14  /*  0- 7 */
	.byte	 4,  0,  8,  0,  0,  0,  0, 15  /*  8-15 */
	.byte	11,  5,  0,  0,  9,  0,  0, 26  /* 16-23 */
	.byte	 0,  0,  0,  0,  0, 22, 28, 16  /* 24-31 */
	.byte	32, 12,  6,  0,  0,  0,  0,  0	/* 32-39 */
	.byte	10,  0,  0, 25,  0,  0, 21, 27  /* 40-47 */
	.byte	31,  0,  0,  0,  0, 24,  0, 20  /* 48-55 */
	.byte   30,  0, 23, 19, 29, 18, 17,  0  /* 56-63 */
#else
	clzne	r0, r0
	rsbne	r0, r0, #32
	RET
#endif
